### PR TITLE
Detect mixed-size and mixed-atomicity non-synchronized accesses

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/miri/issues/2303

### CODE REVIEW

**Review by oli-obk — APPROVED** [review]

File: tests/pass/concurrency/simple.rs [hunk 1]
```diff
@@ -62,6 +62,23 @@ fn panic_named() {
         .unwrap_err();
 }
 
+// THis is not a data race!
```
Comment (taiki-e) [inline tests/pass/concurrency/simple.rs:65]:
```suggestion
// This is not a data race!
```

File: tests/pass-dep/concurrency/linux-futex.rs [hunk 2]
```diff
@@ -267,7 +267,7 @@ fn concurrent_wait_wake() {
         FUTEX.store(FREE, Ordering::Relaxed);
         unsafe {
             DATA = 1;
-            libc::syscall(libc::SYS_futex, &FUTEX as *const AtomicI32, libc::FUTEX_WAKE, 1);
+            libc::syscall(libc::SYS_futex, addr_of!(FUTEX), libc::FUTEX_WAKE, 1);
```
Comment (taiki-e) [inline tests/pass-dep/concurrency/linux-futex.rs:270]: I'm confused about the changes to this file. Does this mean that creating a reference *to an atomic type* is considered a data race?

File: tests/pass/concurrency/simple.rs (hunk 1 above)
Comment (RalfJung) [reply tests/pass/concurrency/simple.rs]: Thanks!

**Pull request merged**